libraries/GCS_MAVLink/GCS.h

```diff
@@ -470,11 +465,80 @@ class GCS_MAVLINK
+    };
+    // returns index of id in deferred_message[] or -1 if not present
+    int8_t deferred_message_offset(const ap_message id) const;
+    // returns index of a message in deferred_message[] which should
+    // be sent (or -1 if none to send at the moment)
+    int8_t deferred_message_to_send_offset();
+    // cache of which deferred message should be sent next:
+    int8_t _next_deferred_message_to_send = -1;
+
+    struct deferred_message_bucket_t {
+        Bitmask ap_message_ids{MSG_LAST};
+        uint16_t interval_ms;
+        uint16_t last_sent_ms; // from AP_HAL::millis16()
+    };
+    deferred_message_bucket_t deferred_message_bucket[5];
+    static const uint8_t no_bucket_to_send = -1;
+    static const ap_message no_message_to_send = (ap_message)-1;
+    uint8_t sending_bucket_id = no_bucket_to_send;
+    Bitmask bucket_message_ids_to_send{MSG_LAST};
+
+    ap_message next_deferred_message_to_send();
+    void find_next_bucket_to_send();
+
+    // bitmask of IDs the code has spontaneously decided it wants to
+    // send out.  Examples include HEARTBEAT (gcs_send_heartbeat)
+    Bitmask pushed_ap_message_ids{MSG_LAST};
+
+    // returns true if it is OK to send a message while we are in
+    // delay callback.  In particular, when we are doing sensor init
+    // we still send heartbeats.
+    bool send_message_in_delay_callback(const ap_message id) const;
+
+    // if true is returned, interval will contain the default interval for id
+    bool default_interval_for_ap_message(const ap_message id, uint16_t &interval) const;
+    //  if true is returned, interval will contain the default interval for id
+    bool default_interval_for_mavlink_message_id(const uint32_t mavlink_message_id, uint16_t &interval) const;
+    // returns an interval in milliseconds for any ap_message in stream id
+    uint16_t interval_for_stream(GCS_MAVLINK::streams id) const;
+    // set an inverval for a specific mavlink message.  Returns false
+    // on failure (typically because there is no mapping from that
+    // mavlink ID to an ap_message)
+    bool set_mavlink_message_id_interval(const uint32_t mavlink_id,
+                                         const uint16_t interval_ms);
+    // map a mavlink ID to an ap_message which, if passed to
+    // try_send_message, will cause a mavlink message with that id to
+    // be emitted.  Returns MSG_LAST if no such mapping exists.
+    ap_message mavlink_id_to_ap_message_id(const uint32_t mavlink_id) const;
+    // set the interval at which an ap_message should be emitted (in ms)
+    bool set_ap_message_interval(enum ap_message id, uint16_t interval_ms);
+    // call set_ap_message_interval for each entry in a stream,
+    // the interval being based on the stream's rate
+    void initialise_message_intervals_for_stream(GCS_MAVLINK::streams id);
+    // call initialise_message_intervals_for_stream on every stream:
+    void initialise_message_intervals_from_streamrates();
```

Comment:
I think if "initialise_message_intervals_for_stream" and "initialise_message_intervals_from_streamrates" really do the same thing then "initialise_message_intervals_from_streamrates" should perhaps be named "initialise_message_intervals_for_all_streams"

Reply:
Well.... initialise_message_intervals_for_all_streams" may not do all of the streams.  I think that's reasonable to be considered internal.  In fact, there are several mavlink streams not initialised.